### PR TITLE
Add proof of non-existence and refactor

### PR DESCRIPTION
**Pull request by cloudhead**
Add non-existence proof and refactor.

### CODE REVIEW

**Comment by ethanfrey** [comment]
Note: this addresses issue #30

File: Makefile [hunk 1]
```diff
@@ -4,9 +4,9 @@ GOTOOLS = \
 
 PDFFLAGS=-pdf --nodefraction=0.1
 
-all: test install
+all: get_vendor_deps install test
```
Comment (ethanfrey) [inline Makefile:7]: :+1:

File: iavl/iavl_proof.go [hunk 2]
```diff
@@ -2,11 +2,12 @@ package iavl
 
 import (
 	"bytes"
+	"fmt"
 
 	"golang.org/x/crypto/ripemd160"
 
-	. "github.com/tendermint/tmlibs/common"
 	"github.com/tendermint/go-wire"
+	. "github.com/tendermint/tmlibs/common"
```
Comment (ethanfrey) [inline iavl/iavl_proof.go]: Note: I know you did not write the module, but  `import .` is deprecated in our codebase.
Proper for tmlibs/common is `cmn` or `common`

File: iavl/iavl_proof2.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+	"github.com/pkg/errors"
+	"github.com/tendermint/go-wire"
+)
+
+type PathToKey struct {
+	LeafHash   []byte
+	InnerNodes []IAVLProofInnerNode
+}
+
+func (p *PathToKey) String() string {
```
Comment (ethanfrey) [inline iavl/iavl_proof2.go]: More readable this way for sure.
Please add a comment on PathToKey to say the last InnerNodes is closest to root, this was confusing.

File: iavl/iavl_proof2.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+			return errors.New("failed to verify right path")
+		}
+		if bytes.Compare(proof.RightNode.KeyBytes, key) != 1 {
+			return errors.New("left node key must be greater than supplied key")
+		}
+	}
+
+	// Both paths exist, check that they are sequential.
+	if proof.RightPath != nil && proof.LeftPath != nil {
+		// TODO
```
Comment (ethanfrey) [inline iavl/iavl_proof2.go]: Note to self: review after finished

File: iavl/iavl_proof2.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+func (proof *KeyNotExistsProof) Verify(key []byte, root []byte) error {
+	if !bytes.Equal(proof.RootHash, root) {
+		return errors.New("roots do not match")
+	}
+
+	if proof.LeftPath == nil && proof.RightPath == nil {
+		return errors.New("at least one path must exist")
+	}
+	if proof.LeftPath != nil {
```
Comment (ethanfrey) [inline iavl/iavl_proof2.go:119]: Very clear, self-explanatory code. :+1:

File: iavl/iavl_proof2.go [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+	RootHash   []byte
+	PathToKeys []*PathToKey
+}
+
+func (proof *KeyRangeExistsProof) Verify(key []byte, value []byte, root []byte) bool {
+	return false
+}
+
+func ReadKeyExistsProof(data []byte) (*KeyExistsProof, error) {
+	proof := new(KeyExistsProof)
```
Comment (ethanfrey) [inline iavl/iavl_proof2.go]: Oh how I love go-wire and **KeyExistsProof....

File: iavl/iavl_proof2.go [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+			Right:  nil,
+		}
+		proof.InnerNodes = append(proof.InnerNodes, branch)
+		return value, true
+	}
+	return nil, false
+}
+
+func (node *IAVLNode) constructKeyNotExistsProof(t *IAVLTree, key []byte, proof *KeyNotExistsProof) error {
+	idx, _, exists := t.Get(key)
```
Comment (ethanfrey) [inline iavl/iavl_proof2.go]: Please add comment that idx is the index of the first key greater than the requested key if it doesn't exist (or max+1 if higher than all)

File: iavl/iavl_proof2.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+		}
+		node.constructKeyExistsProof(t, rkey, rproof)
+		proof.RightPath = &rproof.PathToKey
+		proof.RightNode = IAVLProofLeafNode{KeyBytes: rkey, ValueBytes: rval}
+	}
+
+	return nil
+}
+
+func (t *IAVLTree) ConstructKeyExistsProof(key []byte) (value []byte, proof *KeyExistsProof) {
```
Comment (ethanfrey) [inline iavl/iavl_proof2.go]: Do these IAVLTree.Construct* methods need to be public?  Shall we only expose them through `GetWithProof()`... or whatever you call the method?

I like to keep the public API relatively minimal, like only one method to do a given action.

File: iavl/iavl_test.go [hunk 9]
```diff
@@ -570,6 +584,93 @@ func TestIAVLTreeProof(t *testing.T) {
 	}
 }
 
+func TestIAVLTreeKeyExistsProof(t *testing.T) {
```
Comment (ethanfrey) [inline iavl/iavl_test.go]: Do you want to move this to another test file (iavl_proof_test.go ?) I think it can be nice to separate old and new tests and code, like one file is refactored, the other needs to be....

File: iavl/iavl_test.go [hunk 10]
```diff
@@ -570,6 +584,93 @@ func TestIAVLTreeProof(t *testing.T) {
...
+
+	// insert lots of info and store the bytes
+	keys := make([][]byte, 200)
+	for i := 0; i < 200; i++ {
+		key, value := randstr(20), randstr(200)
+		tree.Set([]byte(key), []byte(value))
+		keys[i] = []byte(key)
+	}
+
+	// query random key fails
```
Comment (ethanfrey) [inline iavl/iavl_test.go]: want to add a test for lower than minimum and higher than maximum?

File: iavl/iavl_tree.go [hunk 11]
```diff
@@ -94,6 +94,15 @@ func (t *IAVLTree) Proof(key []byte) (value []byte, proofBytes []byte, exists bo
 	return value, proofBytes, true
 }
 
+func (t *IAVLTree) KeyExistsProof(key []byte) (value []byte, proofBytes []byte, exists bool) {
```
Comment (ethanfrey) [inline iavl/iavl_tree.go]: Do we need this?  Can we just use `GetWithProof`?

File: iavl/iavl_test.go [hunk 12]
```diff
@@ -570,6 +584,93 @@ func TestIAVLTreeProof(t *testing.T) {
...
+	err = proof.Verify(key, root)
+	require.Nil(err, "%+v", err)
+
+	// Query existing keys fails without proof of non-existence.
+	for _, key := range keys {
+		proof, err := tree.ConstructKeyNotExistsProof(key)
+		require.NotNil(err, "%+v", err)
+		require.Nil(proof)
+	}
+}
```
Comment (ethanfrey) [inline iavl/iavl_test.go]: Can you add a test to produce invalid proofs of non-existence to make sure they are rejected?
Like containing valid left and right proofs, but the combination doesn't actually demonstrate that the given key is missing.
I think you are working on this as we speak however....

**Review by ethanfrey — COMMENTED** [review]
Great work.  A few comments on polishing it up.

Once you finish those last tests (and verify no fake non-existence proofs pass), feel free to collapse the commits (or not) as you prefer. And I will do a final review for merge.

File: iavl/iavl_tree.go [hunk 13]
```diff
@@ -158,6 +167,14 @@ func (t *IAVLTree) GetByIndex(index int) (key []byte, value []byte) {
 	return t.root.getByIndex(t, index)
 }
 
+func (t *IAVLTree) GetWithProof(key []byte) ([]byte, *KeyExistsProof, *KeyNotExistsProof, error) {
+	value, proof := t.ConstructKeyExistsProof(key)
+	if proof == nil {
+		return nil, nil, nil, nil
```
Comment (ethanfrey) [inline iavl/iavl_tree.go]: In this case, we should return the KeyNotExistsProof, correct?

File: iavl/iavl_tree.go (hunk 11 above)
Comment (ethanfrey) [reply iavl/iavl_tree.go]: Yeah, please remove, `GetWithProof` is a nicer interface

File: iavl/iavl_tree.go (hunk 13 above)
Comment (ethanfrey) [reply iavl/iavl_tree.go]: Wait, does this ever return error?  Either have an error case, or remove it from the signature please.

File: iavl/iavl_tree.go (hunk 11 above)
Comment (cloudhead) [reply iavl/iavl_tree.go]: Indeed, this is redundant, will remove.

File: iavl/iavl_test.go [hunk 14]
```diff
@@ -570,6 +584,134 @@ func TestIAVLTreeProof(t *testing.T) {
...
+		exists := false
+
+		for _, k := range keys {
+			if bytes.Compare(key, k) == 0 {
+				exists = true
+				break
+			}
+		}
+
+		if exists {
```
Comment (ethanfrey) [inline iavl/iavl_test.go]: Nice test cases, thorough job.

File: iavl/iavl_proof2.go [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+	proof = &KeyExistsProof{
+		RootHash: t.root.hash,
+	}
+	if value, exists := t.root.constructKeyExistsProof(t, key, proof); exists {
+		return value, proof
+	}
+	return nil, nil
+}
+
+func (t *IAVLTree) ConstructKeyNotExistsProof(key []byte) (*KeyNotExistsProof, error) {
```
Comment (ethanfrey) [inline iavl/iavl_proof2.go]: Also vote for `constructKeyNotExistsProof` unless we really need to expose directly.

File: iavl/iavl_proof2.go [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+			return errors.New("left node key must be greater than supplied key")
+		}
+	}
+
+	// Both paths exist, check that they are sequential.
+	if proof.RightPath != nil && proof.LeftPath != nil {
+		lns := proof.LeftPath.InnerNodes[:]
+		rns := proof.RightPath.InnerNodes[:]
+
+		for bytes.Compare(lns[len(lns)-1].Left, rns[len(rns)-1].Left) == 0 &&
```
Comment (ethanfrey) [inline iavl/iavl_proof2.go]: I think this could be cleaner somehow without all the `lns[len(lns)-1]`... stuff but have no concrete idea how... maybe I'll add one. I know you don't like it either.

Comment (ethanfrey) [reply iavl/iavl_proof2.go]: I have an idea...

```
		var ln, rn IAVLProofInnerNode
		lns := proof.LeftPath.InnerNodes[:]
		rns := proof.RightPath.InnerNodes[:]
		pop := func() {
			ln, lns = lns[len(lns)-1], lns[:len(lns)-1]
			rn, rns = rns[len(rns)-1], rns[:len(rns)-1]
		}

		// keep poping them off until we pop off exactly one that doesn't match
		for pop(); bytes.Equal(ln.Left, rn.Left) && bytes.Equal(ln.Right, rn.Right); pop() {
		}
		lpath := &PathToKey{InnerNodes: lns}
		rpath := &PathToKey{InnerNodes: rns}
```

This works and passes tests, maybe you can be inspired by it, and make it even better
